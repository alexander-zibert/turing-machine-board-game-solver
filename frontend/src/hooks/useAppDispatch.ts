import { useDispatch } from "react-redux";
import type { AppDispatch } from "../store";

// Use throughout your app instead of plain `useDispatch` to get strongly typed dispatch
export const useAppDispatch = () => useDispatch<AppDispatch>();
