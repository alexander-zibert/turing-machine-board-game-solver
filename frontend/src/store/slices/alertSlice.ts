import { AlertColor } from "@mui/material";
import { PayloadAction, createSlice } from "@reduxjs/toolkit";

export type AlertPayload = {
  message: string;
  level: AlertColor;
};

export type AlertState = AlertPayload & {
  open: boolean;
};

const initialState: AlertState = {
  open: false,
  message: "",
  level: "info",
};

export const alertSlice = createSlice({
  name: "alert",
  initialState,
  reducers: {
    load: (_, action: PayloadAction<AlertState>) => action.payload,
    reset: () => initialState,
    openAlert: (state, action: PayloadAction<AlertPayload>) => {
      const { payload } = action;
      state.open = true;
      state.level = payload.level;
      state.message = payload.message;
    },
    closeAlert: (state) => {
      state.open = false;
    },
  },
});

export const alertActions = alertSlice.actions;

export default alertSlice.reducer;
