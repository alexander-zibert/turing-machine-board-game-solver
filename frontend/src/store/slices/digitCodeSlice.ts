import { PayloadAction, createSlice } from "@reduxjs/toolkit";

export type DigitCodeState = {
  shape: Shape;
  digit: Digit;
  state: "correct" | "incorrect";
}[];

const initialState: DigitCodeState = [];

export const digitCodeSlice = createSlice({
  name: "digitCode",
  initialState,
  reducers: {
    load: (_, action: PayloadAction<DigitCodeState>) => action.payload,
    reset: () => initialState,
    toggleDigitState: (
      state,
      action: PayloadAction<{ shape: Shape; digit: Digit }>
    ) => {
      const { shape, digit } = action.payload;

      const index = state.findIndex(
        (entry) => entry.shape === shape && entry.digit === digit
      );

      if (index >= 0) {
        state.splice(index, 1);
      } else {
        state.push({ shape, digit, state: "incorrect" });
      }
    },
    setDigitState: (
      state,
      action: PayloadAction<{ shape: Shape; digit: Digit; isValid: boolean }>
    ) => {
      const { shape, digit, isValid } = action.payload;
      const index = state.findIndex(
        (entry) => entry.shape === shape && entry.digit === digit
      );

      if (index >= 0) {
        state.splice(index, 1);
      }

      if (!isValid) {
         state.push({ shape, digit, state: "incorrect" });
      }
    },
  },
});

export const digitCodeActions = digitCodeSlice.actions;

export default digitCodeSlice.reducer;
