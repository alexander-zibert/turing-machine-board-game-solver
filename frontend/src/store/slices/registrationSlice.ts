import { PayloadAction, createSlice } from "@reduxjs/toolkit";

export type RegistrationState = {
  name: string;
  hash: string;
  status: "new" | "fetch" | "ready";
};

const initialState: RegistrationState = {
  name: "",
  hash: "",
  status: "new",
};

export const registrationSlice = createSlice({
  name: "registration",
  initialState,
  reducers: {
    load: (_, action: PayloadAction<RegistrationState>) => action.payload,
    reset: () => initialState,
    fetch: (state) => ({
      ...state,
      status: "fetch",
    }),
    fetchBad: (state) => ({
      ...state,
      status: "new",
    }),
    fetchDone: (state) => ({
      ...state,
      status: "ready",
    }),
    updateName: (state, action: PayloadAction<string>) => {
      state.name = action.payload;
    },
    updateHash: (state, action: PayloadAction<string>) => {
      state.hash = action.payload
        .replace("#", "")
        .replaceAll(" ", "")
        .split(/(.{3})/)
        .filter(e => e)
        .join(" ");
    },
  },
});

export const registrationActions = registrationSlice.actions;

export default registrationSlice.reducer;
