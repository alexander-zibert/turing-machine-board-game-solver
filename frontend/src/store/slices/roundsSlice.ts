import { PayloadAction, createSlice } from "@reduxjs/toolkit";

type Query = {
  verifier: Verifier;
  state: "solved" | "unsolved" | "unknown";
};

type Code = {
  shape: Shape;
  digit: Nullable<Digit>;
};

export type RoundsState = {
  code: Code[];
  queries: Query[];
  isPristine: boolean;
}[];

const initialState: RoundsState = [
  {
    code: (["triangle", "square", "circle"] as Shape[]).map((shape) => ({
      shape,
      digit: null,
    })),
    queries: (["A", "B", "C", "D", "E", "F"] as Verifier[]).map((verifier) => ({
      verifier,
      state: "unknown",
    })),
    isPristine: false,
  },
];

export const roundsSlice = createSlice({
  name: "rounds",
  initialState,
  reducers: {
    load: (_, action: PayloadAction<RoundsState>) => action.payload,
    reset: () => initialState,
    deleteRound: (state, action: PayloadAction<number>) => {
      state.splice(action.payload, 1);
    },
    addRound: (state) => {
      state.push({
        code: (["triangle", "square", "circle"] as Shape[]).map((shape) => ({
          shape,
          digit: null,
        })),
        queries: (["A", "B", "C", "D", "E", "F"] as Verifier[]).map(
          (verifier) => ({
            verifier,
            state: "unknown",
          }),
        ),
        isPristine: true,
      });
    },
    updateCodeDigit: (
      state,
      action: PayloadAction<{
        index: number;
        shape: Shape;
        digit: Nullable<Digit>;
      }>,
    ) => {
      const { index, shape, digit } = action.payload;
      const round = state[index];
      const code = round.code.find((code) => code.shape === shape)!;

      code.digit = digit;

      round.isPristine = false;

      state[index] = round;
    },
    updateQueryState: (
      state,
      action: PayloadAction<{ index: number; verifier: Verifier }>,
    ) => {
      const { index, verifier } = action.payload;
      const round = state[index];
      const query = round.queries.find((query) => query.verifier === verifier)!;

      switch (query.state) {
        case "unknown":
          query.state = "unsolved";
          break;
        case "unsolved":
          query.state = "solved";
          break;
        case "solved":
          query.state = "unknown";
          break;
      }

      round.isPristine = false;

      state[index] = round;
    },
  },
});

export const roundsActions = roundsSlice.actions;

export default roundsSlice.reducer;
