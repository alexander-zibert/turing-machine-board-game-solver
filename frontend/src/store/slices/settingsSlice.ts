import { createSlice } from "@reduxjs/toolkit";

export type SettingsState = {
  paletteMode: PaletteMode;
  storeVersion: number;
  language: string;
};

const initialState: SettingsState = {
  paletteMode: "light",
  storeVersion: parseInt(process.env.REACT_APP_STORE_VERSION),
  language: "EN",
};

export const settingsSlice = createSlice({
  name: "settings",
  initialState,
  reducers: {
    togglePaletteMode: (state) => {
      state.paletteMode = state.paletteMode === "light" ? "dark" : "light";
    },
    updateLanguage: (state, action) => {
      state.language = action.payload;
    },
  },
});

export const settingsActions = settingsSlice.actions;

export default settingsSlice.reducer;
