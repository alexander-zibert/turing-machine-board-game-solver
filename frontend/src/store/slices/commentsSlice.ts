import { PayloadAction, createSlice } from '@reduxjs/toolkit'
import { CriteriaCard, criteriaCardPool } from 'hooks/useCriteriaCard'

const verifiers: Verifier[] = ['A', 'B', 'C', 'D', 'E', 'F']

const shuffle = (array: number[]) => {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[array[i], array[j]] = [array[j], array[i]]
  }
  return array
}

type Comment = {
  verifier: Verifier
  criteriaCards: CriteriaCard[]
}

export type CommentsState = Comment[]

const initialState: CommentsState = []

export const commentsSlice = createSlice({
  name: 'comments',
  initialState,
  reducers: {
    load: (_, action: PayloadAction<CommentsState>) => action.payload,
    reset: () => initialState,
    setCards: (
      state,
      action: PayloadAction<{
        fake?: number[]
        ind: number[]
        m?: number
      }>
    ) => {
      const { fake, ind, m } = action.payload

      for (let i = 0; i < ind.length; i++) {
        if (fake) {
          const cards = [ind[i], fake[i]]
          const shuffledCards = shuffle(cards)

          state.push({
            verifier: verifiers[i],
            criteriaCards: [
              criteriaCardPool.find(cc => cc.id === shuffledCards[0])!,
              criteriaCardPool.find(cc => cc.id === shuffledCards[1])!,
            ],
          })
        } else {
          const card = ind.sort((n1, n2) => n1 - n2)[i]

          state.push({
            verifier: verifiers[i],
            criteriaCards: [
              {
                ...criteriaCardPool.find(cc => cc.id === card)!,
                nightmare: m === 2,
              },
            ],
          })
        }
      }
    },
    updateCard: (
      state,
      action: PayloadAction<{
        verifier: Verifier
        index: number
        card?: CriteriaCard
      }>
    ) => {
      const { verifier, index, card } = action.payload

      if (!card) return

      const comment = state.find(comment => comment.verifier === verifier)
      if (comment) {
        comment.criteriaCards[index] = card
      } else {
        state.push({
          criteriaCards: [card],
          verifier,
        })
      }
    },
  },
})

export const commentsActions = commentsSlice.actions

export default commentsSlice.reducer
