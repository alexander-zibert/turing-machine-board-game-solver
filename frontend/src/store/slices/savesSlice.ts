import { PayloadAction, createSlice } from '@reduxjs/toolkit'
import { CommentsState } from './commentsSlice'
import { DigitCodeState } from './digitCodeSlice'
import { RegistrationState } from './registrationSlice'
import { RoundsState } from './roundsSlice'

export type Save = {
  comments: CommentsState
  date: number
  digitCode: DigitCodeState
  registration: RegistrationState
  rounds: RoundsState
}

type SavesState = Save[]

const initialState: SavesState = []

export const savesSlice = createSlice({
  name: 'saves',
  initialState,
  reducers: {
    save: (state, action: PayloadAction<Save>) => {
      const { comments, date, digitCode, registration, rounds } = action.payload

      if (registration.hash === '') return state

      const saveIndex = state.findIndex(
        save => save.registration.hash === registration.hash
      )

      if (state[saveIndex]) {
        state[saveIndex].comments = comments
        state[saveIndex].digitCode = digitCode
        state[saveIndex].registration = registration
        state[saveIndex].rounds = rounds
        return state
      } else {
        state.push({
          comments,
          date,
          digitCode,
          registration,
          rounds,
        })
      }
    },
    deleteSave: (state, action: PayloadAction<number>) => {
      const saveIndex = state.findIndex(save => save.date === action.payload)

      state.splice(saveIndex, 1)
    },
  },
})

export const savesActions = savesSlice.actions

export default savesSlice.reducer
