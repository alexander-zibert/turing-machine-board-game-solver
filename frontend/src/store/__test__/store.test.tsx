import { renderHook } from '@testing-library/react'
import { useAppDispatch } from 'hooks/useAppDispatch'
import { PropsWithChildren } from 'react'
import { Provider } from 'react-redux'
import { settingsActions } from 'store/slices/settingsSlice'
import { store } from '..'

describe('store', () => {
  const wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  )

  it('should match initial state', () => {
    expect(store.getState()).toMatchSnapshot()
  })

  it('should toggle palette mode', () => {
    const appDispatch = renderHook(useAppDispatch, { wrapper })

    appDispatch.result.current(settingsActions.togglePaletteMode())

    expect(store.getState()).toMatchSnapshot()
  })
})
