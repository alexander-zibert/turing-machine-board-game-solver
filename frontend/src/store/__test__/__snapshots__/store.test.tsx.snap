// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should match initial state 1`] = `
Object {
  "alert": Object {
    "level": "info",
    "message": "",
    "open": false,
  },
  "comments": Array [],
  "digitCode": Array [],
  "registration": Object {
    "hash": "",
    "name": "",
    "status": "new",
  },
  "rounds": Array [
    Object {
      "code": Array [
        Object {
          "digit": null,
          "shape": "triangle",
        },
        Object {
          "digit": null,
          "shape": "square",
        },
        Object {
          "digit": null,
          "shape": "circle",
        },
      ],
      "isPristine": false,
      "queries": Array [
        Object {
          "state": "unknown",
          "verifier": "A",
        },
        Object {
          "state": "unknown",
          "verifier": "B",
        },
        Object {
          "state": "unknown",
          "verifier": "C",
        },
        Object {
          "state": "unknown",
          "verifier": "D",
        },
        Object {
          "state": "unknown",
          "verifier": "E",
        },
        Object {
          "state": "unknown",
          "verifier": "F",
        },
      ],
    },
  ],
  "saves": Array [],
  "settings": Object {
    "language": "EN",
    "paletteMode": "light",
    "storeVersion": 1695655881142,
  },
}
`;

exports[`store should toggle palette mode 1`] = `
Object {
  "alert": Object {
    "level": "info",
    "message": "",
    "open": false,
  },
  "comments": Array [],
  "digitCode": Array [],
  "registration": Object {
    "hash": "",
    "name": "",
    "status": "new",
  },
  "rounds": Array [
    Object {
      "code": Array [
        Object {
          "digit": null,
          "shape": "triangle",
        },
        Object {
          "digit": null,
          "shape": "square",
        },
        Object {
          "digit": null,
          "shape": "circle",
        },
      ],
      "isPristine": false,
      "queries": Array [
        Object {
          "state": "unknown",
          "verifier": "A",
        },
        Object {
          "state": "unknown",
          "verifier": "B",
        },
        Object {
          "state": "unknown",
          "verifier": "C",
        },
        Object {
          "state": "unknown",
          "verifier": "D",
        },
        Object {
          "state": "unknown",
          "verifier": "E",
        },
        Object {
          "state": "unknown",
          "verifier": "F",
        },
      ],
    },
  ],
  "saves": Array [],
  "settings": Object {
    "language": "EN",
    "paletteMode": "dark",
    "storeVersion": 1695655881142,
  },
}
`;
